### PR TITLE
issue/6378 - media settings caption and link support

### LINKED ISSUE #6378: Aztec: Improve Media Settings
There are some missing fields from the Image Settings screen in wpandroid:

* Alignement
* Link
* Caption

wpios issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/7596

### PR DESCRIPTION
**Pull request by khaykov**
Fixes #6378
Adds ability to edit image link and caption.

I discovered a minor parsing issue on the Aztec side - when you add a caption, a newline is added in front of it, so it would appear below the target image. When converting post to HTML, this newline turn's into `<br>`, and this `<br>` messes up with a post inside the editor and on the web.
Fix removes that newline, resizes caption span to fill the gap <del>and summons the demons of the night</del>. Behavior is also different based on do we have text after caption or not. We should remove it as soon as this issue is fixed on the Aztec side.

screenshot-2017-12-14_17 23 21 899

### CODE REVIEW

**Comment by nbradbury** [comment]
When I enter a caption, it gets wrapped to multiple lines with a large space in the middle (perhaps this is what you're describing above regarding newlines?). After I've added a caption, if I return to the image settings the first letter is missing from the caption I added.

Screenshots below illustrate what I mean.

screenshot_1513256811

**Comment by maxme** [comment]
> When I enter a caption, it gets wrapped to multiple lines with a large space in the middle (perhaps this is what you're describing above regarding newlines?). After I've added a caption, if I return to the image settings the first letter is missing from the caption I added.

@0nko I wonder if this was changed in a recent Aztec update (but not in `v1.0-beta.12`)

**Comment by khaykov** [comment]
This is probably has something to do with workaround for https://github.com/wordpress-mobile/AztecEditor-Android/issues/572
At least dissapearing character.

**Comment by nbradbury** [comment]
@khaykov For reference, here's the HTML that's generated after adding an image with a caption.

screenshot_1513261100

**Comment by khaykov** [comment]
@nbradbury Thanks for details! I added another step that should help with the issue. `CaptionExtension` didn't expect changes from the client side, and this resulted in missing character and render issues when wrapping text. I'm  resetting text, after passing it through parser/formatter in order to normalize it. It adds a really minor performance hit on huge posts, but I think it's good enough for temp workaround.

**Comment by nbradbury** [comment]
@khaykov That seems to have fixed the problems! I did run into another issue, though: if you add a caption then manually remove the image, the image is removed but the caption remains. This doesn't happen if you remove the image from within image settings - it's only when you use backspace in the editor to remove it.

screenshot_1513344257

screenshot_1513344266

**Comment by khaykov** [comment]
Thanks for your patience, Nick! Caption is a neverending story.
It is now removed when you delete the image with backspace. Working on this problem, I found a [bigger issue](https://github.com/wordpress-mobile/AztecEditor-Android/issues/573). I managed to fix part of it on the client side, but one problem remains - when you delete caption text, including newline after the image, caption span expands to the text below it. Unfortunately, this is something that should be fixed on the Aztec side.

File: libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java [hunk 1]
```diff
@@ -1008,6 +1008,35 @@ public void onMediaDeleted(AztecAttributes aztecAttributes) {
         if (!TextUtils.isEmpty(localMediaId)) {
             mEditorFragmentListener.onMediaDeleted(localMediaId);
         }
+        removeCaptionFromDeletedMedia(localMediaId);
```
Comment (nbradbury) [inline libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: I think this should be inside the `if` clause above, so it doesn't get called when `localMediaId` is empty.

Comment (nbradbury) [reply libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: Related to the above, this fix isn't working for me. I added a photo from the WP media library (not one from the device) then added a caption, but when I deleted it the caption remained. Debugging here showed the `localMediaId` was null.

Comment (khaykov) [reply libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/AztecEditorFragment.java]: You are right, this is not going to work reliably in all cases. We can't rely on `localMediaId`, or any id devised from image attributes for this matter. It's also possible to have multiple images with the same id, and we will have no idea what caption to remove. Looks like this is another thing that has to be done on the Aztec side - we need to listen for removed image spans and delete any captions attached to them.

**Comment by nbradbury** [comment]
There are some remaining issues on the Aztec side, but this PR looks good :shipit:

**Pull request merged**